chalk-integration/src/query.rs

```diff
@@ -50,7 +51,7 @@ pub trait LoweringDatabase: RustIrDatabase<ChalkIr> {
     /// cached state becomes invalid, so the query is marked as
     /// volatile, thus ensuring that the solver is recreated in every
     /// revision (i.e., each time source program changes).
-    #[salsa::volatile]
+    #[salsa::dependencies]
```

Comment:
Hmm, I don't know that this has quite the same semantics. I think this is going to create a fresh solver every time. This was an issue that was raised over on the salsa repo -- that we lost a way to have values that are not `Eq`.

Reply:
So is there a way to work around this, or are we stuck waiting for a solution to the salsa issue?

Reply:
> Hmm, I don't know that this has quite the same semantics. I think this is going to create a fresh solver every time. This was an issue that was raised over on the salsa repo -- that we lost a way to have values that are not `Eq`.

Ah yes, you're right. I was hoping they had the same semantics going off [their descriptions](https://salsa-rs.github.io/salsa/rfcs/RFC0001-Query-Group-Traits.html#controlling-query-modes), but looking at the source code they definitely aren't the same.

> So is there a way to work around this, or are we stuck waiting for a solution to the salsa issue?

We could use a hacky workaround (which @nikomatsakis briefly alluded to). I just pushed some changes that *should* work, but it's not ideal.

Reply:
> Hmm, I don't know that this has quite the same semantics. I think this is going to create a fresh solver every time. This was an issue that was raised over on the salsa repo -- that we lost a way to have values that are not `Eq`.

Ah yes, you're right. I was hoping they had the same semantics going off [their descriptions](https://salsa-rs.github.io/salsa/rfcs/RFC0001-Query-Group-Traits.html#controlling-query-modes), but looking at the source code they definitely aren't the same.

> So is there a way to work around this, or are we stuck waiting for a solution to the salsa issue?

We could use a hacky workaround (which @nikomatsakis briefly alluded to). I just pushed some changes that *should* work, but it's not ideal.